### ISSUE TITLE
correspondance between atlas region labels and `NiftiLabelsMasker`'s output dimensions

### ISSUE THREAD

**Issue by jeromedockes** [nilearn/nilearn#3085]
`NiftiLabelsMasker.__init__` takes a discrete atlas (3D map of integers, `labels_img`) and a list of labels (region names, strings, `labels`).
It is assumed that values in the image are region ids between 0 and N, where N is `len(labels)`, and that the region where the image takes the value `j` (0 ≤ j < N) corresponds to the human-readable name `labels[j]`.
Some regions can be empty in the original atlas image or become empty after this atlas image is resampled to the resolution of the data that the masker must transform.
The output of `NiftiLabelsMasker.transform(func_img)` is a vector where each entry contains the average of `func_img` in a given region of the (possibly resampled) atlas image.
Empty regions in the atlas image are discarded: they are not assigned an entry (dimension) in the output of `transform` (we cannot compute the average of 0 voxels).
Moreover, the background (usually the region with value `0` in the atlas image, although that can be controlled) is also discarded.
(For some reason, if a region becomes empty due to computing intersection with the `mask_img` --we can provide a mask in addition to the atlas image, and everything outside the mask is considered as background-- then it still has a dimension in the output but that gets filled with `0` [here](https://github.com/nilearn/nilearn/blob/2fd6665664f27dfbf1f10aeeba660adc9b560361/nilearn/regions/signal_extraction.py#L132), which seems a bit inconsistent.)
Therefore, there is not a direct correspondance between entries of `transform`'s output and entries in `labels` or values in the original atlas image.
This means that `masker.transform(img).T[j]` does not correspond to the region whose id is `j` and whose label is `labels[j]`.

To handle this, the `NiftiLabelsMasker` has a `labels_` attribute that allows to map values in the atlas image (region ids) to dimensions in the output of `transform` (the masked image).
It is a vector of integers of length `M`, where M ≤ N is the number of dimensions in the masked image, and for 0 ≤ j < M, `labels_[j]` gives the corresponding atlas region id. 
That means that the region where the atlas image takes value `masker.labels_[j]` gets averaged to produce `masked_img.T[j]`, and that the corresponding label (region name string) is `labels[labels_[j]]`.

To summarize: the `j`th entry of `masker.transform(img)` corresponds to the region name `masker.labels[masker.labels_[j]]`.
However, this is not documented (and arguably not straightforward, and the name `labels_` is not very explicit and confusingly similar to `labels`).
That causes surprises as users can expect that it would simply be `masker.labels[j]`, and even nilearn itself sometimes gets it wrong, for example:
- when attempting to generate reports [here](https://github.com/nilearn/nilearn/blob/5cba192e3f1183330158f8d7469f36c01ba1f534/nilearn/input_data/nifti_labels_masker.py#L252), which causes the error [reported here](https://github.com/nilearn/nilearn/issues/2036#issuecomment-975007816)
- when giving the impression that we should just add 1 to account for the background in [this example](https://github.com/nilearn/nilearn/blob/2fd6665664f27dfbf1f10aeeba660adc9b560361/examples/06_manipulating_images/plot_nifti_labels_simple.py#L83)

To make it easier to use these maskers, we could:
- document the `labels_` attribute and show the correct way to access the label for a given output dimension
- add utility functions or attributes to help users identify the correct labels, so they don't have to type themselves `masker.labels[masker.labels_[j]]`. Something like `column_names` and `column_region_ids`? or maybe `output_dimension` instead of `column`?

We could also consider:
- being consistent in how we treat empty regions, ie do not allocate an output dimension for regions that are not empty in the `labels_img` but have an empty intersection with `mask_img`. Actually the interface would be simpler if there was only one way to specify what is the background, rather than having both the background label and the `mask_img`, but we cannot change that without breaking backwards compatibility.
- allowing specifying the labels as a dictionary -- there may be valid reasons for not having consecutive integers as region ids, eg storing different levels of a hierarchical atlas in different parts of the binary representation of these region ids

and we can fix the report generation

**Comment by jeromedockes** [comment]
Here is what was decided in the office hours.

for the NiftiLabelsMasker:
- PR 1:
  - improve the documentation of the NiftiLabelsMasker's `labels` parameter to make it clear that `labels[j]` must correspond to the region where `atlas_img` takes value `j`
  - create a clearly documented attribute containing `masker.labels[maskers.labels_]` so users can easily find the correspondance between masked image dimensions and string labels
- PR 2:
  - allow passing `labels` as a dictionary to avoid errors due to this implicit mapping between region ids and region names

for the report generation:
- PR 3:
  - do not raise a ValueError when there are empty regions
- Issue 1:
  - think about generating reports showing the resampled atlas after transform?

for atlases:
- PR 4 (after #3083):
  - create attributes containing an explicit mapping (dictionary) between region ids and region names

**Comment by bthirion** [comment]
Thx ! 
Just a small comment on terminology: I would not use the 'column' term, as it is unclear for users that the components they deal with are actually columns in a given matrix.

**Comment by mtorabi59** [comment]
Hi,
I'd like to work on this issue. So as far as I understood the difference between `masker.labels` and `masker.labels[maskers.labels_]` is only the background label, because the other missing labels will still be present in `masker.labels[maskers.labels_]`. I think the best idea is to remove the missing labels and not allocate a dimension to them in the output.  Then, we can put contents of `masker.labels[maskers.labels_]` which are the labels remained after fitting the masker in `maskers.labels_`. We can also generate a warning about the labels that are being removed. Is there a reason the `maskers.labels_` was kept the way it is?
Moreover, I didn't find a documentation explaining `maskers.labels_` and `maskers.labels`, so as suggested by @jeromedockes I will provide a proper documentation for them which also clarifies their difference. I can make an example for it too.

**Comment by mtorabi59** [comment]
@jeromedockes could you explain your idea for PR2 a bit more?

**Comment by jeromedockes** [comment]
> because the other missing labels will still be present in `masker.labels[maskers.labels_]`

are you sure that is the case? AFAICT only non-empty regions in the mask get dimensions in the output. ie if my atlas has regions 0 (background), 1, 2, 3, but region 2 is empty (no voxels hold the value 2 in the atlas image), the output of the masker will have 2 dimensions (corresponding to regions 1 and 3)

**Comment by jeromedockes** [comment]
> Is there a reason the `maskers.labels_` was kept the way it is?

I think its main problem is its name. it gives, for each output dimension, the index of the corresponding region in the full list of atlas region names. therefore it is not a list of labels but a list of indices in the list of labels.

one problem is that too many things are called labels -- the atlas region names, the atlas region ids (the value that voxels of that region contain in the atlas image), and even indices in the list of atlas region ids. one of the things you could improve when working on the documentation is clarify a bit the terminology and add a little glossary

**Comment by jeromedockes** [comment]
> @jeromedockes could you explain your idea for PR2 a bit more?

ATM if I want to give names to the atlas regions I must pass an array (`labels`) where `labels[i]` is the name for the part of the image where voxels take the value `i`. so if my atlas image contains 0 for background, 12 for left hemisphere and 135 for right hemisphere, I must create a labels array of length 136.  I would like to be able to pass a dictionary like `{"background": 0, "left hemisphere": 12, "right hemisphere": 135}`

**Comment by mtorabi59** [comment]
> > because the other missing labels will still be present in `masker.labels[maskers.labels_]`
> 
> are you sure that is the case? AFAICT only non-empty regions in the mask get dimensions in the output. ie if my atlas has regions 0 (background), 1, 2, 3, but region 2 is empty (no voxels hold the value 2 in the atlas image), the output of the masker will have 2 dimensions (corresponding to regions 1 and 3)

Sorry, I mean the labels of the regions that became empty as a result of no intersection with an additional mask. So as you said I think no dimension should be allocated to them and they shouldn't be in `maskers.labels_`

**Comment by mtorabi59** [comment]
could some one explain this comment [here](https://github.com/nilearn/nilearn/blob/3109b060e4484db6ce9f77acb5880219fc54c9ed/nilearn/regions/signal_extraction.py#L128)? It might be related to the reason that the empty labels are filled with zero rather than being removed.

**Comment by ymzayek** [comment]
> Sorry, I mean the labels of the regions that became empty as a result of no intersection with an additional mask. So as you said I think no dimension should be allocated to them and they shouldn't be in `maskers.labels_`

I agree with this and would make behavior more consistent with treating `0` values. 

> could some one explain this comment [here](https://github.com/nilearn/nilearn/blob/3109b060e4484db6ce9f77acb5880219fc54c9ed/nilearn/regions/signal_extraction.py#L128)? It might be related to the reason that the empty labels are filled with zero rather than being removed.

@mtorabi59 from the assoctiated PR from the git blame it seems to me that it has to do with scipy not being able to handle `nan` values so these are replaced with `0`'s.

**Comment by mtorabi59** [comment]
So I guess a simple change in the order of the commands in this [block](https://github.com/nilearn/nilearn/blob/2fd6665664f27dfbf1f10aeeba660adc9b560361/nilearn/regions/signal_extraction.py#L107) will solve the problem! So, basically we first apply the additional mask and then find the list of all existing labels, instead of first finding them and then setting them to zero. This way they will be treated as background from the begining. And then there is no need for this [block.](https://github.com/nilearn/nilearn/blob/2fd6665664f27dfbf1f10aeeba660adc9b560361/nilearn/regions/signal_extraction.py#L128)



    # Perform computation
        labels_data = _safe_get_data(labels_img, ensure_finite=True)
    
        if mask_img is not None:
            mask_data = _safe_get_data(mask_img, ensure_finite=True)
            labels_data = labels_data.copy()
            labels_data[np.logical_not(mask_data)] = background_label
	    
	    labels = list(np.unique(labels_data))
       if background_label in labels:
            labels.remove(background_label)

**Comment by mtorabi59** [comment]
Then we can solve the confusion between `maskers.labels` and `maskers.labels_` by changing [this line](https://github.com/nilearn/nilearn/blob/f62e896234e389c9579bbf41b3c2531111cc6371/nilearn/maskers/nifti_labels_masker.py#L646):

    - self.labels_ = labels_
    + self.labels_ = [self.labels[label_idx] for label_idx in labels_]

then `maskers.labels_` will include the label names corresponding to columns in `signals`. 

Afterwards, I will provide an updated documentation for these two attributes and explain their difference.

**Comment by jeromedockes** [comment]
thanks @mtorabi59 ! I agree with both points.   for the second one I think it will still be useful to have a mapping from integer values contained in the atlas image to output dimensions, especially in the case where the user did not provide any region labels. sometimes regions don't have names, eg when the "atlas" is the result of clustering

**Comment by jeromedockes** [comment]
As @mtorabi59 and I just discussed orally:

There are 3 important attributes for each atlas region:
- the (optional) region human-readable name, such as "Parietal lobe"
- the region id -- integer value that indicates a voxel in the atlas map belongs to that region. For example, in a given atlas all the voxels that belong to the parietal lobe may contain the value 1, all voxels that belong to the frontal lobe the value 2, etc.
- the index of the corresponding signal in the output of the masker's `transform` method. For example, the masker may output an array of shape (n, k), where n is the number of time points and k the number of non-empty (after masking) regions, and it might put the parietal lobe (corresponding to value 1 in the atlas map) in the first column of that matrix. (the background, which often has region id 0, isn't given a signal or dimension in `transform`'s output). in that case the index for the parietal lobe in the masker output would be 0.

ATM it is a bit confusing for users because everything is called `labels` and the documentation doesn't say how to make these conversions.
For example to go from masker output index to region name (the dashed arrow) we do:
```
REGION_NAME = masker.labels[masker.labels_[INDEX_IN_MASKER_OUTPUT]]
```

We should do a small inventory and for each of the arrows below:
- decide if users might be interested in that conversion
- if yes, how can it be done at the moment
- if there is no convenient way, how do we add it to the API

Ideally come up with very distinct names for these things (not `labels` vs `labels_`), and show how to do the useful conversions in the documentation

![region_attributes](https://user-images.githubusercontent.com/9196501/226055400-668e4fa1-274e-4713-93b3-515f3900408f.svg)

**Comment by bthirion** [comment]
I would call the first one "label", the second one "index" and I don't know what to do with the third one, as it depends on choices done at masking time.
Bzw for me in the above diagram there is no direct arrow btw Region ID and Region name. Their relationship is mediated by the index AFAICT.

**Comment by mtorabi59** [comment]
My idea is that we can add another attribute other than the two that I previously suggested (`region_signals` already exists, and `region_labels`), named `region_maps`. The point is because of the additional mask that user might provide, the original atlas maps might not be useful anymore. So I suggest we introduce another attribute, `region_maps`, that is the final region maps corresponding to the regions in `region_signals` and is in fact the non-empty original regions after applying the additional mask. 

So in summary, after applying transform, three attributes will be provided to the user:

1. `region_signals` : a numpy ndarray of size (n, k), n being number of time points and k being number of non-empty regions, including signals of regions.
2.  `region_labels` : a list of label names (e.g. 'parietal lobe') of length k, corresponding to the k regions in `region_signals`
3.  `region_maps`: a Nifti1Image similar to original atlas maps, but including only the non-empty regions (and excluding the background). Each voxel of a region gets the index value that that region has in `region_signals`.
So for instance, if the first column of `region_signals` , `region_signals[:, 0]`, corresponds to 'parietal lobe', the first item in `region_labels`, `region_labels[0]`, will be 'parietal lobe', and all the voxels that belong to 'parietal lobe' in `region_maps` will have the value `0`. 
To solve the background problem, I suggest that we fill the background and empty voxels in `region_maps` with `np.nan` instead of `0`, which makes more sense, so that we don't have any problem matching the indices between the three attributes.
let me know if you have any questions.

**Comment by bthirion** [comment]
I think that this makes sense.

**Comment by mtorabi59** [comment]
So I realized that the `img_to_signals_maps` also has the same behavior of returning empty regions and for this reason one of the tests fails when comparing the results of `img_to_signals_maps` and `signals_to_img_labels`. I have to solve the issue for `img_to_signals_maps` too so that they give consistent outputs, but maybe first we should discuss if we want to remove empty regions in both functions. Personaly, I still think that it makes sense to do so, because the user probably has to do that at some point anyways, no?

**Comment by bthirion** [comment]
I think that the point was to avoid a multi-step approach to simplify the API.
This is why I still prefer the first solution.

**Comment by ymzayek** [comment]
If we go with the first option, we can introduce the new parameter and deprecate the behavior in 2 release cycles (warning that default behavior will change in release 0.13) and then in release 0.13 deprecate the new parameter and therefore completely removing the new behavior in 1 release cycle.

@bthirion I think the argument for the second option is: why let users give a ROI mask (the label or maps img) and then another mask on top of that to mask the ROI? I understand your comment about avoiding the multi-step approach but why should the user not be expected to provide a "final" mask themselves?

Another thing, it seems to me the second option would give more freedom to the user, but at the cost of having to do it in 2 steps

Also @mtorabi59 after looking at the API more closely, I think we can already do what you suggest in the new function with https://github.com/nilearn/nilearn/blob/9c29e3376243b643a72f78b8b98a56d12c2e99bb/nilearn/masking.py#L118

**Comment by jeromedockes** [comment]
> 3\. To solve the background problem, I suggest that we fill the background and empty voxels in `region_maps` with `np.nan` instead of `0`, which makes more sense, so that we don't have any problem matching the indices between the three attributes.

the atlas (`region_maps`) contains integers, so it cannot store `np.nan`. One option would be to put `-1`. However, I'm not 100% that this is better than keeping `0`, as `0` is used to mean "background" in most atlases, and by default in the NiftiLabelsMasker's interface. Moreover, it might be good to keep the same values as in the original atlas for regions that do not disappear

**Comment by jeromedockes** [comment]
> Also @mtorabi59 after looking at the API more closely, I think we can already do what you suggest in the new function with

I'm not sure that is the case, because AFAICT this will return a mask (an array of booleans), whereas what we would need is to crop the atlas (and return an array of integers)

**Comment by jeromedockes** [comment]
> @bthirion I think the argument for the second option is: why let users give a ROI mask (the label or maps img) and then another mask on top of that to mask the ROI? I understand your comment about avoiding the multi-step approach but why should the user not be expected to provide a "final" mask themselves?

I agree and that is why I thought we should consider the option of removing that parameter. It is a bit awkward to provide both an atlas and a mask, it complicates the API and implementation, and it adds one potentially conflicting affine (users are often understandably confused about what gets resampled to what resolution). Also, is cropping the atlas by far the most frequent atlas transformation -- why offer a `mask_img` and not for example a list of region labels to keep?

However thinking about it a bit more, (i) building the final image used for `transform` (`mask_img_`, `labels_img_`, `maps_img_` depending on the masker) is considered part of the masker's responsibilities (for example `NiftiMasker` can compute a mask from the functional images; `NiftiMapsMasker` also has a `mask_img`) (ii) letting the masker crop the atlas allows it to keep track of the new region labels for the user (iii) removing `mask_img` would be an important backward incompatibility (iv) removing `mask_img` does not remove the possibility of empty regions because those can appear due to resampling (v) `mask_img` adds a bit of convenience if combining a standard atlas with a custom cropping mask is a frequent use case. So I think I also prefer keeping it after all, if we believe that providing `mask_img` is a frequent use case.

**Comment by jeromedockes** [comment]
FWIW the `mask_img` parameter is used in none of the 12 examples that use either `NiftiLabelsMasker` or `NiftiMapsMasker` in the gallery

**Comment by bthirion** [comment]
Indeed we tend to use adult images, and are not much worried about grey/white matter separation. Other users may have a different perspective on this.

**Comment by mtorabi59** [comment]
So in summary, should I go ahead with option one, which is removing empty regions in the output of [img_to_signals_labels](https://github.com/nilearn/nilearn/blob/f62e896234e389c9579bbf41b3c2531111cc6371/nilearn/regions/signal_extraction.py#L21) and [img_to_signals_maps](https://github.com/nilearn/nilearn/blob/f62e896234e389c9579bbf41b3c2531111cc6371/nilearn/regions/signal_extraction.py#L235), through Deprecation cycle? 
And next I can:

1. improve the reporting so that it makes users aware of the result of applying `mask_img` on the atlas.
currently the report looks like something like this (the green line shows the applied `mask_img`):
<img width="1142" alt="image" src="https://user-images.githubusercontent.com/72619172/232626875-46dc7977-889e-430a-9973-5aa50b053213.png">

2. add new attributes besides `region_signals` like `region_maps` and `region_labels`

**Comment by mtorabi59** [comment]
> (iv) removing `mask_img` does not remove the possibility of empty regions because those can appear due to resampling

I don't think resampling will cause empty regions in `region_signals` because the `missing_labels` are found after resampling is already done. The labels passed to `img_to_signals_labels` are only the ones remained after resampling. The regions/labels removed in resampling won't be in `labels_img` anymore, right? So `img_to_signals_labels` is not aware of them to include them in `region_signals`.

**Comment by jeromedockes** [comment]
> So `img_to_signals_labels` is not aware of them to include them in `region_signals`

right -- what I meant is that some regions from the original atlas provided by the user may not have a corresponding column in the masker's output because they disappeared due to resampling

**Comment by jeromedockes** [comment]
> So in summary

yes, that's a good plan! 
plus updating the documentation to remove the confusion due to calling everything `labels`, eg
```
labels list of str, optional
    Full labels corresponding to the labels image. This is used to improve 
    reporting quality if provided. Warning: The labels must be consistent with
    the label values provided through labels_img.
```

should be changed

**Comment by Remi-Gau** [comment]
extra constraint in all this to comply with sklearn estimator checks: transform should not modify the object()

**Comment by Remi-Gau** [comment]
This issue should probably revisited in the light of the support of BIDS look up tables that was added to the label maskers:

Indirectly it sort of addressed

> PR 2:
>     allow passing labels as a dictionary to avoid errors due to this implicit mapping between region ids and region names

although it did introduce a new `lut` parameters instead of the allowing labels to be anything else than a list of strings

**Comment by mtorabi59** [comment]
@Remi-Gau I agree, it addressed PR2 very well.  Also defining the private variable `_region_id_name` was for addressing the same issue.

**Comment by Remi-Gau** [comment]
OK except for the last few bugs related to this in the last release I will consider this one done
